### PR TITLE
Multiple imputation using chained equation (MICE)

### PR DESCRIPTION
**Pull request by Shafaq-Siddiqi**
1. Main DML script (mice_linearReg.dml)
2. Java test file (BuiltinMiceLinearRegTest.java)
3. DML test script (mice_linearRegression.dml)

### CODE REVIEW

File: scripts/builtin/mice_linearReg.dml [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+#
+# Modifications Copyright 2019 Graz University of Technology
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
```
Comment (phaniarnab) [inline scripts/builtin/mice_linearReg.dml]: A modification license is required when modifying an existing file.  Was this algorithm existing before?

File: scripts/builtin/mice_linearReg.dml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+# ---------------------------------------------------------------------------------------------
+# dataset               Double   ---        imputed dataset
+# singleSet             Double   ---        A complete dataset generated though a specific iteration
+
+
+
+m_mice_linearReg = function(Matrix[Double] X, Integer iter = 3, Integer complete = 3)
+return(Matrix[Double] dataset, Matrix[Double] singleSet)
+  {
+  n = nrow(X)
```
Comment (phaniarnab) [inline scripts/builtin/mice_linearReg.dml]: Please fix the indentation of the opening bracket.

File: scripts/builtin/mice_linearReg.dml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  MR = matrix(0, rows=1, cols=col)
+  
+  mu = colMeans(X) #mean of columns for initial imputation
+
+# storing the mask/address of missing values
+  Mask = X==0;
+  Mask_Filled = Mask;
+
+# filling the missing data with thier means
+  X2 = X+(Mask*mu)
```
Comment (phaniarnab) [inline scripts/builtin/mice_linearReg.dml]: You can replace mu with colMeans(X) directly and remove a few unnecessary line spaces throughout.
Please fix the typo in the comment.

File: scripts/builtin/mice_linearReg.dml [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  {
+    for(i in 1:col)
+    {
+	  slice_X = removeEmpty(target=X2, margin = "rows", select = inverseMask[,i])
+	  train_Y = slice_X[,i]
+	  
+	  if(i == 1)
+	  {
+	    train_X = slice_X[, (i+1): col]
+	  }
```
Comment (phaniarnab) [inline scripts/builtin/mice_linearReg.dml]: You may skip the braces for a single line of codes.

File: scripts/builtin/mice_linearReg.dml [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	    train_X= slice_X[, 1:col-1];
+	  }
+	  else{
+	    train_X= cbind(slice_X[, 1:i-1], slice_X[,i+1:col]);
+	  }
+	  
+	  
+# slicing dataset for filling missing values
+	  slice_Y = removeEmpty(target=X2, margin = "rows", select = Mask[,i])
+	  test_Y = slice_Y[,i]
```
Comment (phaniarnab) [inline scripts/builtin/mice_linearReg.dml]: It's a good idea to indent the comment to the same level of the block of code on which it is applied.

File: src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinMiceLinearRegTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import org.tugraz.sysds.test.AutomatedTestBase;
+import org.tugraz.sysds.test.TestConfiguration;
+import org.tugraz.sysds.test.TestUtils;
+
+import java.util.HashMap;
+
+public class BuiltinMiceLinearRegTest extends AutomatedTestBase {
+    private final static String TEST_NAME = "mice_linearRegression";
+    private final static String TEST_DIR = "functions/builtin/";
+    private static final String TEST_CLASS_DIR = TEST_DIR + BuiltinMiceLinearRegTest.class.getSimpleName() + "/";
```
Comment (phaniarnab) [inline src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinMiceLinearRegTest.java]: Please use tabs for indentation for java files.

File: src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinMiceLinearRegTest.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+import org.junit.Test;
+import org.tugraz.sysds.common.Types;
+import org.tugraz.sysds.lops.LopProperties;
+import org.tugraz.sysds.runtime.matrix.data.MatrixValue;
+import org.tugraz.sysds.test.AutomatedTestBase;
+import org.tugraz.sysds.test.TestConfiguration;
+import org.tugraz.sysds.test.TestUtils;
+
+import java.util.HashMap;
```
Comment (phaniarnab) [inline src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinMiceLinearRegTest.java]: Please remove the unused imports and resolve the warnings.

File: src/test/scripts/installDependencies.R [hunk 8]
```diff
@@ -35,4 +35,5 @@ custom_install("outliers");
 custom_install("caret");
 custom_install("Sigmoid");
 custom_install("DescTools");
+custom_install("mice");
```
Comment (phaniarnab) [inline src/test/scripts/installDependencies.R]: Why is this required?

Comment (Shafaq-Siddiqi) [reply src/test/scripts/installDependencies.R]: Thank you for your comments. I have fixed the issue as per your suggestions. I have also tried to fix variable names.

**Comment by Shafaq-Siddiqi** [comment]
Values are imputed back using element-wise operations.

**Comment by mboehm7** [comment]
LGTM - thanks @Shafaq-Siddiqi. I further tried to improve the code a bit via vectorized train/test slicing, vectorized imputation, and a parfor loop over columns as these lm models can be trained independently. Furthermore, the imputation seemed to reuse Mask_Filled across iterations such that later iterations had no effect. 

Now I can also reproduce the issues you mentioned (on spark ops, and without the cut between remove empty operations) and will fix them in subsequent commits.
